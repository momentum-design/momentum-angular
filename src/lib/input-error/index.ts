export * from './input-error.module';
export * from './input-error.component';
