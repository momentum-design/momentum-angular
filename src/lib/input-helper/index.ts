export * from './input-helper.module';
export * from './input-helper.component';
