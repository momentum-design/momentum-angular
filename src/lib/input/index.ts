export * from '../input/input.directive';
export * from '../input/input.service';
export * from '../input-container/input-container.component';
export * from '../input-helper/input-helper.component';
export * from '../input-message/input-message.component';
export * from '../input-section/input-section.component';
export * from './input.module';
