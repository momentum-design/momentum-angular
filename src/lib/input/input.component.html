

<div class="cui-input-group" [ngClass]="wrapperClasses">
  <cui-label *ngIf="label" [label]="label" className="cui-label"></cui-label>

  <input
    *ngIf="!secondaryLabel"
    class="cui-input"
    [ngClass]="inputClasses"
    [(ngModel)]="value"
    [placeholder]="placeholder"
    [readonly]="readOnly"
    [attr.disabled]="disabled ? '' :null"
    [type]="type"
    (blur)="onBlur($event)"
    (focus)="onFocus($event)"
    (keydown)="onKeyDown($event)"
    (mousedown)="onMouseDown($event)"
  />

  <!-- Secondary Label -->
  <div class="cui-label__secondary-label-container" *ngIf="secondaryLabel">
    <input
      class="cui-input"
      [ngClass]="inputClasses"
      [(ngModel)]="value"
      [type]="type"
      [placeholder]="placeholder"
      [readonly]="readOnly"
      [attr.disabled]="disabled ? '' :null"
      (blur)="onBlur($event)"
      (focus)="onFocus($event)"
      (keydown)="onKeyDown($event)"
      (mousedown)="onMouseDown($event)"
      [name]="name"
    />

    <label class="cui-label__secondary-label">
      <span>{{secondaryLabel}}</span>
    </label>
  </div>

  <!-- Error Input -->
  <ng-container *ngIf="errors">
    <cui-input-error
      *ngFor="let error of errors"
      [error]="error"
    >
    </cui-input-error>
  </ng-container>

  <!-- Angular Control Error Validation Message  -->
  <div class="cui-input__messages" role="alert" *ngIf="control.invalid">
    <div class="message">{{errorValid}}</div>
  </div>

  <!-- Helper Text -->
  <cui-input-helper
    *ngIf="inputHelpText"
    [message]="inputHelpText"
  >
  </cui-input-helper>
</div>