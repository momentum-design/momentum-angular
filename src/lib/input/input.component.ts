import { Component, OnInit, Input, OnChanges, SimpleChanges, forwardRef, Output, EventEmitter, Optional, Self} from '@angular/core';
import { NG_VALUE_ACCESSOR, ControlValueAccessor, NgControl } from '@angular/forms';

const cb = () => {};

const CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR: any = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => InputComponent),
  multi: true
};

@Component({
  selector: 'cui-input',
  templateUrl: './input.component.html',
  // providers: [CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR]
})

export class InputComponent implements ControlValueAccessor {
  //internal value model
  private innerValue: any = '';

  private onTouchedCallback: () => void = cb;
  private onChangeCallback: (_: any) => void = cb;

  //get accessor
  get value(): any {
    return this.innerValue;
  };

  set value(v: any) {
    if (v !== this.innerValue) {
      this.innerValue = v;
      this.onChangeCallback(v);
    }
  }

  onBlur(event) {
    this.handleBlur.emit()
    event.stopPropagation();
  }

  onFocus(event) {
    if(this.disabled) {
      event.stopPropagation();
      return;
    }
    this.handleFocus.emit();
  }

  onKeyDown(event){
    this.handleKeyDown.emit();
  }

  onMouseDown(event){
    if(this.disabled) {
      event.stopPropagation();
      return;
    }
    this.handleMouseDown.emit();
  }

  //From ControlValueAccessor interface
  writeValue(value: any) {
    if (value !== this.innerValue) {
      this.innerValue = value;
    }
  }

  //From ControlValueAccessor interface
  registerOnChange(fn: any) {
    this.onChangeCallback = fn;
  }

  //From ControlValueAccessor interface
  registerOnTouched(fn: any) {
    this.onTouchedCallback = fn;
  }

  private determineErrorType = array => {
    return array.reduce((agg, e) => {
      return agg === 'error' ? agg : e.type || '';
    }, '');
  };

  private filterErrorsByType = (array, value) => {
    return array.reduce(
      (agg, e) => (e.type === value ? agg.concat(e.error) : agg),
      []
    );
  };

  get hasRequiredError() {
    return this.control.hasError('required');
  }

  get errorValid() {
    // if (this.control.pristine) {
    //   return '';
    // }
    for(let error in this.errorObj) {
      if (this.control.errors[error]) {
        return this.errorObj[error]
      }
    }
    return '';
  }

  /** @option Optional css class string | ''  */
  @Input() public class: string = '';
  /** @option Sets the disabled attribute of the Input | false */
  @Input() public disabled: boolean = false;
  /** @option Array of objects with error type and error message */
  @Input() public errorArr: any[];
  /** @option Input css class name string */
  @Input() public inputClass: string = '';
  /** @option Help Text to appear under the input | '' */
  @Input() public inputHelpText: string = '';
  /** @option Overall input group size | '' */
  @Input() public inputSize: string = '';
  /** @option Input label text | '' */
  @Input() public label: string = '';
  /** @option Placeholder text to display when Input is empty | '' */
  @Input() public placeholder: string = "";
  /** @option Determines if Input can be edited | false */
  @Input() public readOnly: boolean = false;
  /** @option Secondary Input label | ''  */
  @Input() public secondaryLabel: string = '';
  /** @option Input color theme | '' */
  @Input() public theme: string;
  /** @option Input type | 'text' */
  @Input() public type: string = "text";
  /** @option Sets the attribute name to the input element | '' */
  @Input() public name: string = '';
  /** @option Optional error messages object with angular validators | {} */
  @Input() public errorObj: object = {};

  /** @option function when clicked outside of input */
  @Output() handleBlur: EventEmitter<any> = new EventEmitter();
  /** @option function when input is focused */
  @Output() handleFocus: EventEmitter<any> = new EventEmitter();
  /** @option function when key down on input */
  @Output() handleKeyDown: EventEmitter<any> = new EventEmitter();
  /** @option function when mouse down on input */
  @Output() handleMouseDown: EventEmitter<any> = new EventEmitter();

  public errorType;
  public errors;

  constructor(@Optional() @Self() public control: NgControl) {
    if (this.control) {
      this.control.valueAccessor = this;
    }
  }

  ngOnInit() { }

  ngOnChanges(changes: SimpleChanges){
    if(changes.errorArr){
      let changedArr = changes.errorArr.currentValue

      if(changedArr.length > 0){
        this.errorType = this.determineErrorType(changedArr);
        this.errors = this.errorType && this.filterErrorsByType(changedArr, this.errorType);
      }
    }
  }

  get wrapperClasses() {
    return {
      [this.inputSize + ' columns']: this.inputSize,
      'read-only': this.readOnly,
      'disabled': this.disabled,
      'error': this.errorObj && this.control.invalid,
      ['cui-input-group--' + this.theme]: this.theme,
      [this.errorType]: this.errorType,
      [this.class]: this.class
    };
  }

  get inputClasses() {
    return {
      [this.inputClass]: this.inputClass,
      'read-only': this.readOnly,
      'disabled': this.disabled,
      'dirty': this.value
    };
  }
}

/**
* @component input
* @section default
* @angular
*

<div>
  <cui-input
    [(ngModel)]="dataModel"
    inputSize="small-5"
    label="Default Input"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section error
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    label="Error (Error) Input"
    [errorArr]="[{error: 'This is where the error message would be.', type: 'error'}]"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section warning
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    label="Error (Warning) Input"
    [errorArr]="[{error: 'This is where the success message would be.', type: 'warning'}]"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section success
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    label="Error (Success) Input"
    [errorArr]="[{error: 'This is where the success message would be.', type: 'success'}]"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section disabled
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    [disabled]="true"
    label="Disabled Input"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section read-only
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    [readOnly]="true"
    label="Read Only Input"
    [(ngModel)]="dataModel"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section help-text
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    inputHelpText="Help Text"
    label="Help Text Input"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section secondary-label
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    secondaryLabel="Secondary Label"
  >
  </cui-input>
</div>
*/


