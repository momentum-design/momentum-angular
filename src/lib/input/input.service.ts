import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';

@Injectable()
export class InputService {

  private addBeforePadding = new Subject();
  addBeforePadding$ = this.addBeforePadding.asObservable();

  private addAfterPadding = new Subject();
  addAfterPadding$ = this.addAfterPadding.asObservable();

  onPaddingChange(position: string, value: boolean) {
    this[`add${position}Padding`].next(value);
  }
}

