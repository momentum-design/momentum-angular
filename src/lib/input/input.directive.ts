/** @component input */

import {
  DoCheck,
  Directive,
  ElementRef,
  HostBinding,
  Input,
  OnChanges,
  OnDestroy,
  OnInit,
  Optional,
  Self,
  SimpleChanges,
} from '@angular/core';
import { AutofillMonitor } from '@angular/cdk/text-field';
import { takeUntil } from 'rxjs/operators';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { getSupportedInputTypes, Platform } from '@angular/cdk/platform';
import { InputGroupService } from '../input-group';
import { InputService } from './input.service';
import { Subject } from 'rxjs';
import {
  ControlValueAccessor,
  FormControl,
  FormGroupDirective,
  NgControl,
  NgForm,
} from '@angular/forms';

let nextUniqueId = 0;

// Invalid input type. Using one of these will throw an MDInputUnsupportedTypeError.
const MD_INPUT_INVALID_TYPES = [
  'button',
  'checkbox',
  'file',
  'hidden',
  'image',
  'radio',
  'range',
  'reset',
  'submit'
];

function getMDInputUnsupportedTypeError(type: string): Error {
  return Error(`Input type "${type}" isn't supported by mdInput.`);
}

@Directive({
  selector: `input[mdInput], textarea[mdInput]`,
  exportAs: 'mdInput',
  host: {
    '[attr.id]': 'id',
    '[attr.placeholder]': 'placeholder',
    '[attr.aria-describedby]': '_ariaDescribedby || null',
    '[disabled]': 'disabled',
    '[required]': 'required',
    '[readonly]': 'readonly || null',
    '(blur)': 'handleFocusChange(false)',
    '(focus)': 'handleFocusChange(true)',
    '(input)': 'handleInput()'
  },
  providers: [
    {provide: InputGroupService, useExisting: InputDirective}
  ]
})

export class InputDirective implements ControlValueAccessor, DoCheck, InputGroupService<any>, OnChanges, OnDestroy, OnInit {
  private _destroyed = new Subject<void>();
  private _inputValueAccessor: any;
  private _ngControl: any;
  private _parentForm: any;
  private _parentFormGroup: any;
  private _readonly = false;
  protected _disabled = false;
  protected _id: string;
  protected _previousNativeValue: any;
  protected _required = false;
  protected _type = 'text';
  protected _uid = `md-input-${nextUniqueId++}`;
  readonly stateChanges: Subject<void> = new Subject<void>();
  autofilled = false;
  _ariaDescribedby: string;
  _inputAfter = false;
  _inputBefore = false;
  focused: boolean = false;
  errorState: boolean = false;
  status: string;

  /*** @prop Unique HTML ID used for tying label to HTML input | 'md-in-<uniqueId>' */
  @Input()
  get id(): string { return this._id; }
  set id(value: string) { this._id = value || this._uid; }
  @Input()
  /*** @prop Sets the disabled attribute of the Input | false */
  get disabled(): boolean {
    if (this.ngControl && this.ngControl.disabled !== null) {
      return this.ngControl.disabled;
    }
    return this._disabled;
  }
  set disabled(value: boolean) {
    this._disabled = coerceBooleanProperty(value);

    // Browsers may not fire the blur event if the input is disabled too quickly.
    // Reset from here to ensure that the element doesn't become stuck.
    if (this.focused) {
      this.focused = false;
      this.stateChanges.next();
    }
  }
  /*** @prop Placeholder text to display when Input is empty | '' */
  @Input() public placeholder: string = '';
  /*** @prop Determines if Input can be edited | false */
  @Input()
  get readonly(): boolean { return this._readonly; }
  set readonly(value: boolean) { this._readonly = coerceBooleanProperty(value); }
  /*** @prop Determines if Input is required | false */
  @Input()
  get required(): boolean { return this._required; }
  set required(value: boolean) { this._required = coerceBooleanProperty(value); }
  /** @prop Input shape property | '' */
  @Input() public shape: string = '';
  @Input()
  get type(): string { return this._type; }
  set type(value: string) {
    this._type = value || 'text';
    this._validateType();

    // When using Angular inputs, developers are no longer able to set the properties on the native
    // input element. To ensure that bindings for `type` work, we need to sync the setter
    // with the native property. Textarea elements don't support the type property or attribute.
    if (!this._isTextarea() && getSupportedInputTypes().has(this._type)) {
      (this._elementRef.nativeElement as HTMLInputElement).type = this._type;
    }
  }
  /*** @prop Input value | '' */
  @Input()
  get value(): string {
    return this._inputValueAccessor.value;
   }
  set value(value: string) {
    if (value !== this.value) {
      this._inputValueAccessor.value = value;
      this.stateChanges.next();
    }
  }

  protected _neverEmptyInputTypes = [
    'date',
    'datetime',
    'datetime-local',
    'month',
    'time',
    'week'
  ].filter(t => getSupportedInputTypes().has(t));

  @HostBinding('class') get _class(): string {
    return (
      'md-input' +
      `${this._isTextarea() ? ' md-input--multiline' : ''}` +
      `${this.shape ? ` md-input--${this.shape}` : ''}` +
      `${this._inputBefore ? ` md-input--before` : ''}` +
      `${this._inputAfter ? ` md-input--after` : ''}` +
      `${this.focused ? ` md-active` : ''}` +
      `${this.readonly ? ' md-read-only' : ''}` +
      `${this.disabled ? ' md-disabled' : ''}` +
      `${(this._ngControl && this._ngControl.dirty)  ? ` md-dirty` : ''}`
    );
  }

  constructor(
    protected _elementRef: ElementRef<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement>,
    protected _platform: Platform,
    private _autofillMonitor: AutofillMonitor,
    private inputService: InputService,
    @Optional() @Self() public ngControl: NgControl,
    @Optional() _parentForm: NgForm,
    @Optional() _parentFormGroup: FormGroupDirective
  ) {

    const element = this._elementRef.nativeElement;

    this.id = this.id;
    this._inputValueAccessor = (ngControl && ngControl.valueAccessor) || element;
    this._parentForm = _parentForm;
    this._parentFormGroup = _parentFormGroup;
    this._ngControl = ngControl;

    this.inputService.addBeforePadding$
      .pipe(takeUntil(this._destroyed))
      .subscribe(value => this._inputBefore = coerceBooleanProperty(value));

    this.inputService.addAfterPadding$
      .pipe(takeUntil(this._destroyed))
      .subscribe(value => this._inputAfter = coerceBooleanProperty(value));
  }

  ngDoCheck() {
    if (this.ngControl) {
      // We need to re-evaluate this on every change detection cycle, because there are some
      // error triggers that we can't subscribe to (e.g. parent form submissions). This means
      // that whatever logic is in here has to be super lean or we risk destroying the performance.
      this.updateErrorState();
    }
    // We need to dirty-check the native element's value, because there are some cases where
    // we won't be notified when it changes (e.g. the consumer isn't using forms or they're
    // updating the value using `emitEvent: false`).
    this._dirtyCheckNativeValue();
  }

  ngOnChanges(changes: SimpleChanges): void {
    this.stateChanges.next();
  }

  ngOnInit() {
    if (this._platform.isBrowser) {
      this._autofillMonitor.monitor(this._elementRef.nativeElement).subscribe(event => {
        this.autofilled = event.isAutofilled;
        this.stateChanges.next();
      });
    }
  }

  ngOnDestroy() {
    this._destroyed.next();
    this._destroyed.complete();
    this.stateChanges.complete();

    if (this._platform.isBrowser) {
      this._autofillMonitor.stopMonitoring(this._elementRef.nativeElement);
    }
  }

  protected _isNeverEmpty() {
    return this._neverEmptyInputTypes.indexOf(this._type) > -1;
  }

  /** Checks whether the input is invalid based on the native validation. */
  protected _isBadInput() {
    // The `validity` property won't be present on platform-server.
    const validity = (this._elementRef.nativeElement as HTMLInputElement).validity;
    return validity && validity.badInput;
  }

  protected _isTextarea() {
    return this._elementRef.nativeElement.nodeName.toLowerCase() === 'textarea';
  }

  protected _validateType() {
    if (MD_INPUT_INVALID_TYPES.indexOf(this._type) > -1) {
      throw getMDInputUnsupportedTypeError(this._type);
    }
  }

  get empty(): boolean {
    return !this._isNeverEmpty() && !this._elementRef.nativeElement.value && !this._isBadInput() &&
        !this.autofilled;
  }

  /** Callback for the cases where the focused state of the input changes. */
  handleFocusChange(isFocused: boolean) {
    if (isFocused !== this.focused && (!this.readonly || !isFocused)) {
      this.focused = isFocused;
      this.stateChanges.next();
    }
  }

  handleInput() {
    // This is a noop function and is used to let Angular know whenever the value changes.
    // Angular will run a new change detection each time the `input` event has been dispatched.
    // It's necessary that Angular recognizes the value change, because when floatingLabel
    // is set to false and Angular forms aren't used, the placeholder won't recognize the
    // value changes and will not disappear.
    // Listening to the input event wouldn't be necessary when the input is using the
    // FormsModule or ReactiveFormsModule, because Angular forms also listens to input events.
  }

  onChange: any = () => {};
  onTouch: any = () => {};

  registerOnChange(fn: any): void {
    this.onChange = fn;
  }

  registerOnTouched(fn: any): void {
    this.onTouch = fn;
  }

  setDescribedByIds(ids: string[]) {
    this._ariaDescribedby = ids.join(' ');
  }

  setDisabledState?(isDisabled: boolean): void {
    this.disabled = isDisabled;
  }

  updateErrorState() {
    const oldState = this.errorState;
    const parent = this._parentFormGroup || this._parentForm;
    const control = this.ngControl ? this.ngControl.control as FormControl : null;
    const newState = !!(
      control &&
      control.invalid &&
      (control.dirty || (parent && parent.submitted))
    );

    if (newState !== oldState) {
      this.errorState = newState;
      this.stateChanges.next();
    }
    return;
  }

  writeValue(value: any) {
    if (value !== this._inputValueAccessor.value) {
      this._inputValueAccessor.value = value;
    }
  }

  private _dirtyCheckNativeValue() {
    const newValue = this._elementRef.nativeElement.value;

    if (this._previousNativeValue !== newValue) {
      this._previousNativeValue = newValue;
      this.stateChanges.next();
    }
  }
}

