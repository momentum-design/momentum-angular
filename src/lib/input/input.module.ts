import { CommonModule } from '@angular/common';
import { CUSTOM_ELEMENTS_SCHEMA, ModuleWithProviders, NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { IconModule } from '../icon/icon.module';
import { InputContainerModule } from '../input-container/input-container.module';
import { InputHelperModule } from '../input-helper/input-helper.module';
import { InputMessageModule } from '../input-message/input-message.module';
import { InputSectionModule } from '../input-section/input-section.module';
import { InputDirective } from '../input/input.directive';
import { InputService } from '../input/input.service';
import { LabelModule } from '../label/label.module';

@NgModule({
  declarations: [InputDirective],
  imports: [
    CommonModule,
    FormsModule,
    IconModule,
    InputHelperModule,
    InputContainerModule,
    InputMessageModule,
    InputSectionModule,
    LabelModule,
    ReactiveFormsModule,
  ],
  exports: [
    InputDirective,
    IconModule,
    InputHelperModule,
    InputContainerModule,
    InputMessageModule,
    InputSectionModule,
    LabelModule,
  ],
  providers: [InputService],
  schemas: [CUSTOM_ELEMENTS_SCHEMA],
})
export class InputModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: InputModule,
      providers: [InputService],
    };
  }
}
