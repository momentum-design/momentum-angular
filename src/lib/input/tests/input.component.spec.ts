import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { SimpleChange } from '@angular/core';
import { FormsModule, FormControl } from '@angular/forms';
import { BrowserModule, By } from '@angular/platform-browser';
import { InputComponent } from '../input.component';
import { LabelModule } from '../../label';
import { InputErrorModule } from '../../input-error';
import { InputHelperModule } from '../../input-helper';

describe('InputComponent', () => {
  let testComponent: InputComponent;
  let testFixture: ComponentFixture<InputComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [InputComponent],
      imports: [FormsModule, LabelModule, BrowserModule, InputErrorModule, InputHelperModule],
    }).compileComponents();
  }));

  beforeEach(() => {
    testFixture = TestBed.createComponent(InputComponent);
    testComponent = testFixture.componentInstance;
    (testComponent as any).control = new FormControl();
    testFixture.detectChanges();
  });

  it('should create', () => {
    expect(testComponent).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(testFixture).toMatchSnapshot();
  });

  it('should render a label to the input and check to see correct label Text', () => {
    testComponent.label = 'Test Label';
    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const span = inputNativeElement.querySelector('span');

    expect(span).not.toBeNull();
    expect(span.textContent).toMatch('Test Label');
  });

  it('should render a help text message to the input', () => {
    testComponent.inputHelpText = 'Help Text';
    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const helpText = inputNativeElement.querySelector('p');

    expect(helpText.className).toContain('cui-input__help-text');
  });

  it('should render a secondary label to the input', () => {
    testComponent.secondaryLabel = 'Secondary Label';
    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const secondaryLabel = inputNativeElement.querySelector('.cui-label__secondary-label');

    expect(secondaryLabel.className).toContain('cui-label__secondary-label');
  });

  it('should render a input with password type', () => {
    testComponent.type = 'password';
    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const inputType = inputNativeElement.querySelector('input');

    expect(inputType.type).toBe('password');
  });

  it('should render a input with small-5 size', () => {
    testComponent.inputSize = 'small-5';

    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const size = inputNativeElement.querySelector('.cui-input-group');

    expect(size.className).toContain('small-5');
  });

  it('should render a disabled input', () => {
    testComponent.disabled = true;

    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const inputDisabled = inputNativeElement.querySelector('input');

    expect(inputDisabled.disabled).toBe(true);
  });

  it('should render a readonly input', () => {
    testComponent.readOnly = true;

    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const inputReadOnly = inputNativeElement.querySelector('input');

    expect(inputReadOnly.readOnly).toBe(true);
  });

  it('should render a error text input', () => {
    testComponent.errorArr = [{ error: 'This is where the error message would be.', type: 'error' }];

    testComponent.ngOnChanges({
      errorArr: new SimpleChange(null, testComponent.errorArr, true),
    });

    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const inputError = inputNativeElement.querySelector('.cui-input-group');

    expect(inputError.className).toContain('error');
  });

  it('should render a warning text input', () => {
    testComponent.errorArr = [{ error: 'This is where the warning message would be.', type: 'warning' }];

    testComponent.ngOnChanges({
      errorArr: new SimpleChange(null, testComponent.errorArr, true),
    });

    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const inputError = inputNativeElement.querySelector('.cui-input-group');

    expect(inputError.className).toContain('warning');
  });

  it('should render a success text input', () => {
    testComponent.errorArr = [{ error: 'This is where the success message would be.', type: 'success' }];

    testComponent.ngOnChanges({
      errorArr: new SimpleChange(null, testComponent.errorArr, true),
    });

    testFixture.detectChanges();

    const inputNativeElement = testFixture.nativeElement;
    const inputError = inputNativeElement.querySelector('.cui-input-group');

    expect(inputError.className).toContain('success');
  });
});
