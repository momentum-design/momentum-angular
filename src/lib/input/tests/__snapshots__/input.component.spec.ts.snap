// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputComponent should match snapshot 1`] = `
<cui-input
  class=""
  control={[Function FormControl]}
  determineErrorType={[Function Function]}
  disabled="false"
  errorObj={[Function Object]}
  filterErrorsByType={[Function Function]}
  handleBlur={[Function EventEmitter]}
  handleFocus={[Function EventEmitter]}
  handleKeyDown={[Function EventEmitter]}
  handleMouseDown={[Function EventEmitter]}
  innerValue=""
  inputClass=""
  inputHelpText=""
  inputSize=""
  label=""
  name=""
  onChangeCallback={[Function Function]}
  onTouchedCallback={[Function Function]}
  placeholder=""
  readOnly="false"
  secondaryLabel=""
  type={[Function String]}
>
  <div
    class="cui-input-group"
    ng-reflect-klass="cui-input-group"
    ng-reflect-ng-class="[object Object]"
  >
    
    
    <input
      class="cui-input ng-untouched ng-pristine ng-valid"
      ng-reflect-klass="cui-input"
      ng-reflect-model=""
      ng-reflect-ng-class="[object Object]"
      placeholder=""
      type="text"
    />
    
    
    
    
  </div>
</cui-input>
`;
